lib/pos_network.go

```diff
@@ -0,0 +1,557 @@
+	if err != nil {
+		return false
+	}
+
+	otherEncodedBytes, err := other.ToBytes()
+	if err != nil {
+		return false
+	}
+
+	return bytes.Equal(sigEncodedBytes, otherEncodedBytes)
+}
+
+func (sig *AggregatedBLSSignature) ToBytes() ([]byte, error) {
+	retBytes := []byte{}
+
+	// SignersList
+	if sig.SignersList == nil {
+		return nil, errors.New("AggregatedBLSSignature.ToBytes: SignersList must not be nil")
+	}
+	retBytes = append(retBytes, EncodeBitset(sig.SignersList)...)
+
+	// Signature
+	if sig.Signature == nil {
+		return nil, errors.New("AggregatedBLSSignature.ToBytes: Signature must not be nil")
+	}
+	retBytes = append(retBytes, EncodeBLSSignature(sig.Signature)...)
+
+	return retBytes, nil
+}
+
+func (sig *AggregatedBLSSignature) FromBytes(rr io.Reader) error {
+	var err error
+
+	sig.SignersList, err = DecodeBitset(rr)
+	if err != nil {
+		return errors.Wrapf(err, "AggregatedBLSSignature.FromBytes: Error decoding SignersList")
+	}
+
+	sig.Signature, err = DecodeBLSSignature(rr)
+	if err != nil {
+		return errors.Wrapf(err, "AggregatedBLSSignature.FromBytes: Error decoding Signature")
+	}
+
+	return nil
+}
+
+// TimeoutAggregateQuorumCertificate is an aggregation of timeout messages from 2/3rds
+// of all validators, weighted by stake, which indicates that these validators want to
+// time out a particular view.
+//
+// When validators want to time out a view, they send their high QCs to the block proposer
+// who builds an aggregate QC extending the chain from the highest QC that it received
+// from all validators who timed out. To prove that it has selected the highest QC, the
+// proposer also includes a list of the high QC views that each validator has sent.
+type TimeoutAggregateQuorumCertificate struct {
+
+	// The view that the block proposers has produced a timeout QC for.
+	TimedOutView uint64
+
+	// This is the highest QC that the block proposer received from any validator who
+	// has timed out for the current view.
+	ValidatorsHighQC *QuorumCertificate
+
+	// Here we include a list of the HighQC.View values we got from each of the
+	// validators in the ValidatorsTimeoutHighQCViews field. In addition, for each
+	// unique HighQC.View value we received, we combine all the partial signatures
+	// for that HighQC.View into a single BLSMultiSignature.
+	//
+	//
+	// The aggregated signature is made up of partial signatures for all present
+	// validators, each of whom signed a payload with the pair
+	// (current view, the validator's local HighQC.View).
+	//
+	// The ordering of high QC views and validators in the aggregate signature will
+	// match the ordering of active validators in descending order of stake for the
+	// timed out view's epoch. I.e. index 0 will correspond to the highest-staked active
+	// validator in the epoch, index 1 will correspond to the second-highest-staked active
+	// validator, ...
+	ValidatorsTimeoutHighQCViews         []uint64
+	ValidatorsTimeoutAggregatedSignature *AggregatedBLSSignature
```

Comment:
I think there may have been a misunderstanding of this field. Let's suppose that we have the following setup:
- Validator_01: (TimedOutView = 11, HighQCView = 10)
- Validator_02: (TimedOutView = 11, HighQCView = 10)
- Validator_03: (TimedOutView = 11, HighQCView = 9)
- Validator_04: (TimedOutView = 11, HighQCView = 9)
- Validator_05: (TimedOutView = 11, HighQCView = 8)

If this is the case, then I think we need three aggregated signatures, one for each (TimedOutView, HighQCView) pair. And in the TimeoutAggregateQC, it would look more like:
- ValidatorsUniqueTimeoutHighQCViews []uint64 // <-- Notice Unique added to the name. This would have [8, 9, 10]
- ValidatorsTimeoutAggregatedSignatures []*AggregatedBLSSignature // <-- Notice it's a list of AggregatedBLSSignature. It would look as follows: [Signature((11, 8)), Signature([(11, 9), (11, 9)]), Signature([(11, 10), (11, 10)])]

Let me know if I'm missing something with regard to this. The key is that AggregatedBLSSignature can only sign a single payload, which means that we need a different aggregate signature for each HighQCView that validators have.

Reply:
Link looks broken. Just go to fast_hotstuff_bls.go#L195 in the research code.

Reply:
BLS natively supports signature aggregation for multiple public keys, each with a unique payload. So the schema as defined in the code will actually work. 

See PR https://github.com/deso-protocol/core/pull/575 for concrete examples with unit tests that do multi-payload signature aggregation & verification.

Reply:
Making this as resolved. I've merged PR https://github.com/deso-protocol/core/pull/575 which supports BLS signature aggregation and verification on multiple payloads.